File: 36.valid-sudoku.js
/*
 * @lc app=leetcode id=36 lang=javascript
 *
 * [36] Valid Sudoku
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  let row = board[0].length,
    col = board.length;

  for (let i = 0; i < col; i++) {
    let hashRow = {},
      hashCol = {},
      hashBox = {};
    for (let j = 0; j < row; j++) {
        
      const currentRow = board[i][j];
      const currentCol = board[j][i];
      const box =board[3 * Math.floor(i / 3) + Math.floor(j / 3)][3 * (i % 3) + (j % 3)];
      
      const existingInRow = hashRow[currentRow];
      const existingInCol = hashCol[currentCol];
      const existingInBox = hashBox[box];
      
      if (existingInRow || existingInCol || existingInBox) return false;
      
      if (currentRow !== '.') hashRow[currentRow] = 1;
      if (currentCol !== '.') hashCol[currentCol] = 1;
      if (box !== '.') hashBox[box] = 1;
    }
  }
  return true;
};


// @lc code=end

File: 36.valid-sudoku.cpp
/*
 * @lc app=leetcode id=36 lang=cpp
 *
 * [36] Valid Sudoku
 */

// @lc code=start
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {

        unordered_map<int, int> rowMap[9];
        unordered_map<int, int> colMap[9];
        unordered_map<int, int> boxMap[9];

        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                char current = board[row][col];
                if (current == '.')
                    continue;

                int boxIndex = (row / 3) * 3 + col / 3;

                
                if (rowMap[row][current] || colMap[col][current] ||
                    boxMap[boxIndex][current]) {
                    return false;
                }

        
                rowMap[row][current] = 1;
                colMap[col][current] = 1;
                boxMap[boxIndex][current] = 1;
            }
        }
        return true;
    };
};
// @lc code=end

